### ISSUE TITLE
GetOrAllocateThreadPoolBoundHandle is too slow in certain use case

### ISSUE THREAD

**Issue by ElectricVampire** [dotnet/runtime#108363]
### Description

**UseCase**
I am trying to add oracle support in query plan [visualizer visual studio extension](https://github.com/Giorgi/EFCore.Visualizer/tree/main?tab=readme-ov-file#oracle). The issue I am facing is that [connection.Open()](https://github.com/Giorgi/EFCore.Visualizer/blob/main/src/IQueryableObjectSource/DatabaseProvider.cs#L19) is taking 7 seconds which is timing out the debug visualizer.
Exact same connection.Open take milliseconds when done in without visualizer and I am able to reproduce this all the time.

In following image you can see for slower use case `ConnectIterate` takes 7 seconds which seems to be simple `BeginConnect` call on `TcpClient`
![image](https://github.com/user-attachments/assets/f8c5c466-a83c-43d4-9113-d4d55ea291d9)

I also enabled .Net Diagnostics events which shows where those missing 7 seconds went, between slow and fast .net traces, I can see in slower case GetOrAllocateThreadPoolBoundHandle took 7 Seconds which I think went for `ThreadPoolBoundHandle.BindHandle(this)` but why its not executed in faster use case and why its executed in slower use case.

```
03:45:03.1876310[ConnectStart] address: InterNetwork:16:{5,241,127,0,0,1,0,0,0,0,0,0,0,0}
03:45:03.1880221[Info] thisOrContextObject: SafeSocketHandle#14421545, memberName: GetOrAllocateThreadPoolBoundHandle, message: calling ThreadPool.BindHandle()
03:45:10.7995167[Connected] localEp: IPEndPoint#922213090, remoteEp: (null), socketHash: 33163964
```
**Problem:**
When `connection.Open` happens from VS debugger it times out because it has 5 second limit and for every connection.Open it takes 7 seconds because of `GetOrAllocateThreadPoolBoundHandle`. Need a solution by which this can be reduced.

### Reproduction Steps

We should be able to reproduced it all the time. Just doing connection.Open inside your code is very past but done [via visualizer]([connection.Open()](https://github.com/Giorgi/EFCore.Visualizer/blob/main/src/IQueryableObjectSource/DatabaseProvider.cs#L19)), it's very slow even when its the same process.
I have attached .net trace and oracle trace. Oracle trace can provide you details on thread level information.

[fast_net.txt](https://github.com/user-attachments/files/17177587/fast_net.txt)
[fast_oracle.txt](https://github.com/user-attachments/files/17177588/fast_oracle.txt)
[slow_net.txt](https://github.com/user-attachments/files/17177589/slow_net.txt)
[slow_oracle.txt](https://github.com/user-attachments/files/17177590/slow_oracle.txt)

 

### Expected behavior

GetOrAllocateThreadPoolBoundHandle should return immediately as it did in fast use case

### Actual behavior

GetOrAllocateThreadPoolBoundHandle takes 7 seconds

### Known Workarounds

Nope

**Comment by MihaZupan** [comment]
> 03:45:03.1876310[ConnectStart]
> 03:45:03.1880221[Info] message: calling ThreadPool.BindHandle()
> 03:45:10.7995167[Connected]

> I can see in slower case GetOrAllocateThreadPoolBoundHandle took 7 Seconds

That's not what those logs are implying. These are merely timestamps at different places in the code, so all you can glean from them is that you've reached this line of code at some point:
https://github.com/dotnet/runtime/blob/ba9b3ba83603d1f54c00c45a87344568d4277966/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Windows.cs#L49
and then 7 seconds later you got a connection, not that that specific call took 7 seconds.

You should be able to see how long the connection attempt took outside of .NET itself if you collect a network capture (e.g. Wireshark).

**Comment by ElectricVampire** [comment]
> > 03:45:03.1876310[ConnectStart]
> > 03:45:03.1880221[Info] message: calling ThreadPool.BindHandle()
> > 03:45:10.7995167[Connected]
> 
> > I can see in slower case GetOrAllocateThreadPoolBoundHandle took 7 Seconds
> 
> That's not what those logs are implying. These are merely timestamps at different places in the code, so all you can glean from them is that you've reached this line of code at some point:
> 
> https://github.com/dotnet/runtime/blob/ba9b3ba83603d1f54c00c45a87344568d4277966/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Windows.cs#L49
> 
> 
> and then 7 seconds later you got a connection, not that that specific call took 7 seconds.
> You should be able to see how long the connection attempt took outside of .NET itself if you collect a network capture (e.g. Wireshark).

@MihaZupan 
Thanks for guidance. I took the traces between slow and fast use case. 
[fast.pcapng.txt](https://github.com/user-attachments/files/17178895/fast.pcapng.txt)
[slow.pcapng.txt](https://github.com/user-attachments/files/17178896/slow.pcapng.txt)

**Comment by wfurt** [comment]
so the second capture has RST as the fist packet. It seems like there was some previous connection? And do you connect by name or IP address? It may be worth of trying to create small repro e.g. remove all the unnecessary layers to simplify the investigation.

**Comment by ElectricVampire** [comment]
> so the second capture has RST as the fist packet. It seems like there was some previous connection? And do you connect by name or IP address? It may be worth of trying to create small repro e.g. remove all the unnecessary layers to simplify the investigation.

I tested with both(same result), in this particular example its IP. It will be difficult to break down all the layers of code as its within Oracle and I don't have source code
I also created a ticket with Oracle for this for which a bug is being created, hopefully they can better investigate this.
https://github.com/oracle/dotnet-db-samples/issues/412

**Comment by ScotMac** [comment]
Yes, the RST is is for a totally different connection, as seen by the different source port.

However, the rest of the wireshark traces combined with the ODP traces show this to be a .NET problem, where there is an 8 second delay from TcpClient.Connect. ie:

* Call TcpClient.BeginConnect
   => 5.688: SYN sent :
   <= 5.688: SYN, ACK
   => 5.688: ACK
****** 8 second delay ******
* Return from WaitOne
=> 13.607 Send NS connect packet

ODP trace shows:

2024-10-04 21:37:09.896510 TID:13  (NET)       (CID1) Trying (address=147.154.27.228) (port=1522)
2024-10-04 21:37:17.403137 TID:13  (NET) (EXT) (CID1) TcpTransportAdapter.ConnectIterate()
 
Note, that "(NET) (EXT) (CID1) TcpTransportAdapter.ConnectIterate()" trace happens immediately after the return from the TcpClient.BeginConnect/WaitOne, and therefore should have a timestamp close to the "Trying (address= ..." timestamp. The fact that it is 7 seconds later means control was not returned from TcpClient.Connect/WaitOne to ODP.

 ie, if the issue was an ODP code problem, then the timestamp for "(NET) (EXT) (CID1) TcpTransportAdapter.ConnectIterate()" would be not long after the "Trying (address" and we would see a LATER delay in the traces from later ODP statements.

Specifically, it appears after the tcp 3-way handshake, TcpClient does not return control to ODP for around 8 seconds (in the wireshark trace), and 7 seconds in the ODP trace.

**Comment by ElectricVampire** [comment]
Any update on this?
